command/agent/consul/service_client.go

```diff
@@ -248,9 +248,74 @@ func sidecarTagsDifferent(parent, wanted, sidecar []string) bool {
 	return tagsDifferent(wanted, sidecar)
 }
 
+// proxyUpstreamsDifferent determines if the sidecar_service.proxy.upstreams
+// configurations are different between the desired sidecar service state, and
+// the actual sidecar service state currently registered in Consul.
+func proxyUpstreamsDifferent(wanted *api.AgentServiceConnect, sidecar *api.AgentServiceConnectProxyConfig) bool {
+	// There is similar code that already does this in Nomad's API package,
+	// however here we are operating on Consul API package structs, and they do not
+	// provide such helper functions.
+
+	getProxyUpstreams := func(pc *api.AgentServiceConnectProxyConfig) []api.Upstream {
+		switch {
+		case pc == nil:
+			return nil
+		case len(pc.Upstreams) == 0:
+			return nil
+		default:
+			return pc.Upstreams
+		}
+	}
+
+	getConnectUpstreams := func(sc *api.AgentServiceConnect) []api.Upstream {
+		switch {
+		case sc.SidecarService.Proxy == nil:
+			return nil
+		case len(sc.SidecarService.Proxy.Upstreams) == 0:
+			return nil
+		default:
+			return sc.SidecarService.Proxy.Upstreams
+		}
+	}
+
+	upstreamsDifferent := func(a, b []api.Upstream) bool {
+		if len(a) != len(b) {
+			return true
+		}
+
+		for i := 0; i < len(a); i++ {
+			A := a[i]
+			B := b[i]
+			switch {
+			case A.Datacenter != B.Datacenter:
+				return true
+			case A.DestinationName != B.DestinationName:
+				return true
+			case A.LocalBindAddress != B.LocalBindAddress:
+				return true
+			case A.LocalBindPort != B.LocalBindPort:
+				return true
+			case A.MeshGateway.Mode != B.MeshGateway.Mode:
+				return true
+			case !reflect.DeepEqual(A.Config, B.Config):
```

Comment:
It looks like this big switch/case is a performance optimization to avoid doing `reflect.DeepEqual` on the `api.Upstream` if we can avoid it, but from a maintenance perspective this seems like it's brittle. If we upgrade the Consul library and they add new fields this check will silently break. 

(Alternately, is there maybe a way we can write a unit test that automatically gives us non-default values for all fields so that we have a breaking test on upgrade? I don't know of one off the top of my head but that would be really cool.)

Reply:
I think  you just gave me an excuse to play around with https://blog.golang.org/fuzz-beta :slightly_smiling_face: 

Using `reflect.DeepEqual` is similarly dangerous, for example it always returns false if a function is added to the struct fields:
https://play.golang.org/p/Yfu4d2RdTKd
Really thought the only _correct_ thing to do here would be fore Consul to provide equality methods in their `api` package - they are the ones who define what the thing is and know how it compares.

Reply:
Totally agreed there.